astropy/io/fits/hdu/table.py

```diff
@@ -1138,15 +1139,15 @@ def _dump_data(self, fileobj):
         def format_value(val, format):
             if format[0] == 'S':
                 itemsize = int(format[1:])
-                return '%-*s' % (itemsize, val)
+                return '{:{size}}'.format(val, size=itemsize)
             elif format in np.typecodes['AllInteger']:
                 # output integer
-                return '%21d' % val
+                return '{:21d}'.format(val)
             elif format in np.typecodes['Complex']:
-                return '%21.15g+%.15gj' % (val.real, val.imag)
+                return '{:21.15g}+{:.15g}j'.format(val.real, val.imag)
             elif format in np.typecodes['Float']:
                 # output floating point
-                return '%#21.15g' % val
+                return '{:21.15g}'.format(val)
```

Comment:
Some code snippets to clarify what @bsipocz explained above.

In Python 2:

```python
>>> '%#21.15g' % 12.0
'     12.0000000000000'
>>> '{:#21.15g}'.format(12.0)
ValueError: Alternate form (#) not allowed in float format specifier
>>> '{:21.15g}'.format(12.0)
'                   12'
```

In Python 3:

```python
>>> '%#21.15g' % 12.0
'     12.0000000000000'
>>> '{:#21.15g}'.format(12.0)
'     12.0000000000000'
>>> '{:21.15g}'.format(12.0)
'                   12'
```